### ISSUE TITLE
Angular Elements should keep an HTML attribute value in sync with a property value

### ISSUE THREAD

**Issue by Ahmtk** [angular/angular#23557]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Performance issue
[X] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

We encountered a bug/missing feature in Angular Elements in synchronizing between the property and the HTML attribute that reference to the same variable. 

In other words, when the HTML attribute is changed using setAttribute() method, it will change the property value. However, changing the property value will not affect the HTML attribute value (please see the example below).

**Component Declaration**
```
@Component({
  selector: 'test-component',
  templateUrl: './test-component.component.html',
  styleUrls: ['./test-component.component.scss']
})
export class TestComponent implements OnInit {
  
  @Input('custom-name')
  public customName : string = "test some string";

  ngOnInit() {
  }
}
```
**Component Usage**
```
// Changing Attribute value --> affects the corresponding property
document.querySelector('test-component').setAttribute('custom-name', 'from attribute');
console.log(document.querySelector('test-component).customName); // "from attribute"

// Changing Property value --> DOSE NOT affect the corresponding attribute
document.querySelector('test-component).customName = 'from property';
console.log(document.querySelector('test-component').getAttribute('custom-name')); //"from attribute"
```

## Expected behavior

Properties and HTML Attributes should still in sync, regardless how they were changed.

**E.g.**
```
// Changing Attribute value --> affects the corresponding property
document.querySelector('test-component').setAttribute('custom-name', 'from attribute');
console.log(document.querySelector('test-component).customName); // "from attribute"

// Changing Property value --> affects the corresponding attribute
document.querySelector('test-component).customName = 'from property';
console.log(document.querySelector('test-component').getAttribute('custom-name')); //"from property"
```
Another Subject that related to the same context: Polymer creates automatically for each property an event that has the name "property-name-changed" that will be triggered, when the property is changed. Are you planning to do the same in Angular Elements?

**Comment by kassimir** [comment]
Isn't this just how Javascript works? https://plnkr.co/edit/P3OyifAX9w5fjaUNEd7F?p=preview

**Comment by robwormald** [comment]
This is (for now) a conscious decision - you can easily do this yourself using a `@HostBinding('attr.name') ` 

In the future, we'll look at providing a more integrated solution, but we wanted to be narrow in scope in the API we mark as stable :)

**Comment by Ahmtk** [comment]
@robwormald Thanks for the tip about `@HostBinding('attr.name')`. It is exactly what we need. :)

We will use this logic only for primitive types. For objects (complex types) we are planning to use only properties without HTML attributes. So is there a way to define a web component property that has no corresponding attribute? 

We can live without this feature in the meanwhile. However it will improve the coherence of the component interfaces built using Angular Elements.

**Comment by collinstevens** [comment]
@robwormald I've been experiencing a similar issue trying to use @angular/elements as reusable micro-frontends. I've made a demo project at https://github.com/collinstevens/web-component-sender/tree/host-binding-attr. 

Using the @HostBinding('attr.name') as you suggested, I expected the component to receive updates from it. The attribute changes in the DOM in both the web component and the angular component, but the angular component and the web component are not detecting the changes to the @HostBinding. 

I also have a related question posted here: https://stackoverflow.com/questions/53009894/one-way-from-data-source-to-view-target-data-binding-for-angular-elements

**Comment by andershessellund** [comment]
I think it would be great if angular elements followed the best practices regarding property/attribute syncing outlined here: https://developers.google.com/web/fundamentals/web-components/best-practices

TLDR from the above link is that primitive properties should sync with attributes, rich properties should be property only.

**Comment by danieljancar** [comment]
Hey everyone, is there any update on the feature request status? 🦖
